File: Processview_android/app/src/main/java/com/im/daeseong/processview/CProcessInfo.java
package com.im.daeseong.processview;

import android.annotation.SuppressLint;
import android.app.ActivityManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Environment;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.channels.FileChannel;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class CProcessInfo {

    //실행중인 서비스 목록
    public static List<String> getRunningServicesName(Context context) {
        try {
            List<String> serviceInfo = new ArrayList<>();
            ActivityManager activityManager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);

            for (ActivityManager.RunningServiceInfo srvInfo : activityManager.getRunningServices(Integer.MAX_VALUE)) {
                if (!srvInfo.service.getPackageName().equals("com.im.daeseong.processview")) {
                    serviceInfo.add(srvInfo.service.getPackageName());
                }
            }
            return serviceInfo;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    //최상위 액티비티
    public static List<String> getRunningTopPackageName(Context context) {
        try {
            List<String> topPackageInfo = new ArrayList<>();
            ActivityManager activityManager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);

            for (ActivityManager.RunningTaskInfo activityInfo : activityManager.getRunningTasks(Integer.MAX_VALUE)) {
                if (!activityInfo.topActivity.getPackageName().equals("com.im.daeseong.processview")) {
                    topPackageInfo.add(activityInfo.topActivity.getPackageName());
                }
            }
            return topPackageInfo;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    //최근 실행한 Task 목록
    public static List<String> getCurrentTaskPackageName(Context context) {
        try {
            List<String> currentPackageInfo = new ArrayList<>();
            ActivityManager activityManager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);

            for (ActivityManager.RecentTaskInfo appInfo : activityManager.getRecentTasks(Integer.MAX_VALUE, ActivityManager.RECENT_IGNORE_UNAVAILABLE | ActivityManager.RECENT_WITH_EXCLUDED)) {
                if (!appInfo.baseIntent.getComponent().getPackageName().equals("com.im.daeseong.processview")) {
                    currentPackageInfo.add(appInfo.baseIntent.getComponent().getPackageName());
                }
            }
            return currentPackageInfo;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    //전체앱 목록
    public static List<String> getAllPackageNames(Context context) {
        try {
            List<String> allPackageInfo = new ArrayList<>();
            PackageManager packageManager = context.getPackageManager();

            for (ApplicationInfo applicationInfo : packageManager.getInstalledApplications(PackageManager.GET_META_DATA)) {
                if (!applicationInfo.packageName.equals("com.im.daeseong.processview")) {
                    allPackageInfo.add(applicationInfo.packageName);
                }
            }
            return allPackageInfo;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static boolean copyApk(Context context, String sPackageName) {
        try {
            PackageManager packageManager = context.getPackageManager();
            PackageInfo pkgInfo = packageManager.getPackageInfo(sPackageName, 0);

            String sPublicSourceDir = pkgInfo.applicationInfo.publicSourceDir;
            String sFileName = String.format("%s.apk", sPackageName);

            String sPath = String.format("%s/apkList", Environment.getExternalStorageDirectory().getAbsolutePath());
            if (!directoryExists(sPath)) {
                createDirectory(sPath);
            }

            String fullFilename = String.format("%s/%s", sPath, sFileName);
            copyFile(sPublicSourceDir, fullFilename);

            return true;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return false;
    }

    public static boolean killApp(Context context, String sPackageName) {
        try {
            ActivityManager activityManager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
            activityManager.killBackgroundProcesses(sPackageName);

            PackageManager packageManager = context.getPackageManager();
            ApplicationInfo applicationInfo = packageManager.getApplicationInfo(sPackageName, PackageManager.GET_META_DATA);
            android.os.Process.sendSignal(applicationInfo.uid, android.os.Process.SIGNAL_KILL);

            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    public static boolean install(Context context, String sAppPath) {
        try {
            File file = new File(sAppPath);
            if (file.exists()) {
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setDataAndType(Uri.fromFile(file), "application/vnd.android.package-archive");
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                context.startActivity(intent);
            }
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    public static boolean uninstall(Context context, String sPackageName) {
        try {
            Intent intent = new Intent(Intent.ACTION_DELETE);
            Uri packageURI = Uri.parse("package:" + sPackageName);
            intent.setData(packageURI);
            context.startActivity(intent);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    public static boolean startApp(Context context, String sPackageName) {
        try {
            PackageManager packageManager = context.getPackageManager();
            Intent intent = packageManager.getLaunchIntentForPackage(sPackageName);
            if (intent != null) {
                context.startActivity(intent);
            }
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    public static String getPackageLabel(Context context, String sPackageName) {
        try {
            PackageManager packageManager = context.getPackageManager();
            ApplicationInfo applicationInfo = packageManager.getApplicationInfo(sPackageName, PackageManager.GET_META_DATA);
            return applicationInfo.loadLabel(packageManager).toString();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "";
    }

    public static Drawable getPackageIcon(Context context, String sPackageName) {
        try {
            PackageManager packageManager = context.getPackageManager();
            ApplicationInfo applicationInfo = packageManager.getApplicationInfo(sPackageName, PackageManager.GET_META_DATA);
            return applicationInfo.loadIcon(packageManager);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static PackageItem getPackageInfo(Context context, String sPackageName) {
        try {
            PackageManager packageManager = context.getPackageManager();
            PackageInfo pkgInfo = packageManager.getPackageInfo(sPackageName, 0);
            if (pkgInfo == null) return null;

            SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
            Date first = new Date(pkgInfo.firstInstallTime);
            Date last = new Date(pkgInfo.lastUpdateTime);
            String sInstallTime = format.format(first);
            String sUpdateTime = format.format(last);

            String sPermission = pkgInfo.applicationInfo.permission;
            String sPublicSourceDir = pkgInfo.applicationInfo.publicSourceDir;
            String sFilesize = getFileSize(sPublicSourceDir);

            return new PackageItem(sPackageName, sInstallTime, sUpdateTime, pkgInfo.versionName, sFilesize, sPermission, sPublicSourceDir);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static String getPackageCache(String sPackageName, boolean bClear) {
        String sAbsolutePath = Environment.getExternalStorageDirectory().getAbsolutePath();
        String sPath = String.format("%s/Android/data/%s/cache", sAbsolutePath, sPackageName);

        // Clear cache files
        if (bClear) {
            clearCacheFiles(sPath);
        }

        return getCacheFileSize(sPath);
    }

    private static void clearCacheFiles(String sPath) {
        try {
            File directory = new File(sPath);
            if (directory.isDirectory()) {
                File[] files = directory.listFiles();
                if (files != null) {
                    for (File f : files) {
                        f.delete();
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static String getCacheFileSize(String sPath) {
        try {
            long lFileSize = 0;
            File directory = new File(sPath);
            if (directory.isDirectory()) {
                File[] files = directory.listFiles();
                if (files != null) {
                    for (File f : files) {
                        lFileSize += f.length();
                    }
                }
            }
            return byte2FitMemorySize(lFileSize);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "0kb";
    }

    private static String getFileSize(String sPath) {
        try {
            File file = new File(sPath);
            long len = file.length();
            return byte2FitMemorySize(len);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "";
    }

    @SuppressLint("DefaultLocale")
    private static String byte2FitMemorySize(final long byteNum) {
        if (byteNum < 0) {
            return "shouldn't be less than zero!";
        } else if (byteNum < 1024) {
            return String.format("%.3fB", (double) byteNum);
        } else if (byteNum < 1048576) {
            return String.format("%.3fKB", (double) byteNum / 1024);
        } else if (byteNum < 1073741824) {
            return String.format("%.3fMB", (double) byteNum / 1048576);
        } else {
            return String.format("%.3fGB", (double) byteNum / 1073741824);
        }
    }

    private static void createDirectory(String directoryPath) {
        File file = new File(directoryPath);
        if (!file.exists()) {
            file.mkdirs();
        }
    }

    private static boolean directoryExists(String directoryPath) {
        File file = new File(directoryPath);
        return file.isDirectory();
    }

    private static void copyFile(String sourcePath, String destinationPath) {
        FileChannel inChannel = null;
        FileChannel outChannel = null;
        try {
            File sFile = new File(sourcePath);
            File dFile = new File(destinationPath);

            inChannel = new FileInputStream(sFile).getChannel();
            outChannel = new FileOutputStream(dFile).getChannel();
            inChannel.transferTo(0, inChannel.size(), outChannel);
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (inChannel != null) {
                    inChannel.close();
                }
                if (outChannel != null) {
                    outChannel.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    private static void deleteFile(String filePath) {
        File file = new File(filePath);
        if (file.exists()) {
            file.delete();
        }
    }

    private static boolean fileExists(String filePath) {
        File file = new File(filePath);
        return file.exists();
    }
}

File: processview(kotlin)/app/src/main/java/com/daeseong/processview/CProcessInfo.kt
package com.daeseong.processview

import android.annotation.SuppressLint
import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Environment
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.nio.channels.FileChannel
import java.text.SimpleDateFormat
import java.util.*

object CProcessInfo {

    // 실행 중인 서비스 목록
    fun getRunningServicesName(context: Context): List<String>? {
        try {
            val serviceInfo = mutableListOf<String>()
            val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager

            for (srvInfo in activityManager.getRunningServices(Short.MAX_VALUE.toInt())) {
                if (srvInfo.service.packageName == "com.im.daeseong.processview") {
                    continue
                }
                serviceInfo.add(srvInfo.service.packageName)
            }
            return serviceInfo

        } catch (e: Exception) {
        }

        return null
    }

    // 최상위 액티비티
    fun getRunningTopPackageName(context: Context): List<String>? {
        try {
            val topPackageInfo = mutableListOf<String>()
            val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager

            for (activityInfo in activityManager.getRunningTasks(Short.MAX_VALUE.toInt())) {
                if (activityInfo.topActivity!!.packageName == "com.im.daeseong.processview") {
                    continue
                }
                topPackageInfo.add(activityInfo.topActivity!!.packageName)
            }
            return topPackageInfo

        } catch (e: Exception) {
        }

        return null
    }

    // 최근 실행한 Task 목록
    fun getCurrentTaskPackageName(context: Context): List<String>? {
        try {
            val currentPackageInfo = mutableListOf<String>()
            val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager

            for (appInfo in activityManager.getRecentTasks(Short.MAX_VALUE.toInt(),ActivityManager.RECENT_IGNORE_UNAVAILABLE or ActivityManager.RECENT_WITH_EXCLUDED)) {
                if (appInfo.baseIntent.component!!.packageName == "com.im.daeseong.processview") {
                    continue
                }
                currentPackageInfo.add(appInfo.baseIntent.component!!.packageName)
            }
            return currentPackageInfo

        } catch (e: Exception) {
        }

        return null
    }

    // 전체 앱 목록
    fun getAllPackageNames(context: Context): List<String>? {
        try {
            val allPackageInfo = mutableListOf<String>()
            val packageManager = context.packageManager

            for (applicationInfo in packageManager.getInstalledApplications(PackageManager.GET_META_DATA)) {
                if (applicationInfo.packageName == "com.im.daeseong.processview") {
                    continue
                }
                allPackageInfo.add(applicationInfo.packageName)
            }
            return allPackageInfo

        } catch (e: Exception) {
        }
        return null
    }

    fun copyApk(context: Context, sPackageName: String): Boolean {
        try {
            val packageManager = context.packageManager
            val pkgInfo = packageManager.getPackageInfo(sPackageName, 0)

            val sPublicSourceDir = pkgInfo.applicationInfo.publicSourceDir
            val sFileName = String.format("%s.apk", sPackageName)

            val sPath = String.format("%s/apkList", Environment.getExternalStorageDirectory().absolutePath)
            if (!directoryExists(sPath)) createDirectory(sPath)

            val fullFilename = String.format("%s/%s", sPath, sFileName)
            copyFile(sPublicSourceDir, fullFilename)

            return true

        } catch (e: PackageManager.NameNotFoundException) {
        }
        return false
    }

    fun killApp(context: Context, sPackageName: String): Boolean {
        try {
            val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
            activityManager.killBackgroundProcesses(sPackageName)

            val packageManager = context.packageManager
            val applicationInfo = packageManager.getApplicationInfo(sPackageName, PackageManager.GET_META_DATA)
            android.os.Process.sendSignal(applicationInfo.uid, android.os.Process.SIGNAL_KILL)

            return true
        } catch (e: Exception) {
        }
        return false
    }

    fun install(context: Context, sAppPath: String): Boolean {
        try {
            val file = File(sAppPath)
            if (file.exists()) {
                val intent = Intent(Intent.ACTION_VIEW)
                intent.setDataAndType(Uri.fromFile(file), "application/vnd.android.package-archive")
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                context.startActivity(intent)
            }
            return true

        } catch (e: Exception) {
        }
        return false
    }

    fun uninstall(context: Context, sPackageName: String): Boolean {
        try {
            val intent = Intent(Intent.ACTION_DELETE)
            val packageURI = Uri.parse("package:$sPackageName")
            intent.data = packageURI
            context.startActivity(intent)
            return true

        } catch (e: Exception) {
        }
        return false
    }

    fun startApp(context: Context, sPackageName: String): Boolean {
        try {
            val packageManager = context.packageManager
            val intent = packageManager.getLaunchIntentForPackage(sPackageName)
            if (intent != null)
                context.startActivity(intent)
            return true
        } catch (e: Exception) {
        }
        return false
    }

    fun getPackageLabel(context: Context, sPackageName: String): String {
        try {
            val packageManager = context.packageManager
            val applicationInfo = packageManager.getApplicationInfo(sPackageName, PackageManager.GET_META_DATA)
            return applicationInfo.loadLabel(packageManager).toString()

        } catch (e: Exception) {
        }
        return ""
    }

    fun getPackageIcon(context: Context, sPackageName: String): Drawable? {
        try {
            val packageManager = context.packageManager
            val applicationInfo = packageManager.getApplicationInfo(sPackageName, PackageManager.GET_META_DATA)
            return applicationInfo.loadIcon(packageManager)

        } catch (e: Exception) {
        }
        return null
    }

    fun getPackageInfo(context: Context, sPackageName: String): PackageItem? {
        try {
            val packageManager = context.packageManager
            val pkgInfo = packageManager.getPackageInfo(sPackageName, 0) ?: return null

            val format = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
            val first = Date(pkgInfo.firstInstallTime)
            val last = Date(pkgInfo.lastUpdateTime)
            val sInstallTime = format.format(first)
            val sUpdateTime = format.format(last)

            val sPermission = pkgInfo.applicationInfo.permission
            val sPublicSourceDir = pkgInfo.applicationInfo.publicSourceDir
            val sFilesize = getFileSize(sPublicSourceDir)

            return PackageItem(
                pkgInfo.packageName,
                sInstallTime,
                sUpdateTime,
                pkgInfo.versionName,
                sFilesize,
                sPermission,
                sPublicSourceDir)

        } catch (e: Exception) {
        }
        return null
    }

    fun getPackageCache(sPackageName: String, bClear: Boolean): String {
        val sAbsolutePath = Environment.getExternalStorageDirectory().absolutePath
        val sPath = String.format("%s/Android/data/%s/cache", sAbsolutePath, sPackageName)

        // cache file clear
        if (bClear) clearCacheFiles(sPath)

        return getCacheFileSize(sPath)
    }

    private fun clearCacheFiles(sPath: String) {
        try {
            val directory = File(sPath)
            if (directory.isDirectory) {
                val files = directory.listFiles()
                for (f in files) {
                    f.delete()
                }
            }
        } catch (e: Exception) {
        }
    }

    private fun getCacheFileSize(sPath: String): String {
        try {
            var sFilesize = ""
            var lFileSize: Long = 0
            val directory = File(sPath)
            if (directory.isDirectory) {
                val files = directory.listFiles()
                for (f in files) {
                    lFileSize += f.length()
                }
            }
            sFilesize = byte2FitMemorySize(lFileSize)
            return sFilesize

        } catch (e: Exception) {
        }
        return "0kb"
    }

    private fun getFileSize(sPath: String): String {
        try {
            val file = File(sPath)
            val len = file.length()
            return byte2FitMemorySize(len)

        } catch (e: Exception) {
        }
        return ""
    }

    @SuppressLint("DefaultLocale")
    private fun byte2FitMemorySize(byteNum: Long): String {
        return when {
            byteNum < 0 -> "shouldn't be less than zero!"
            byteNum < 1024 -> String.format("%.3fB", byteNum.toDouble())
            byteNum < 1048576 -> String.format("%.3fKB", byteNum.toDouble() / 1024)
            byteNum < 1073741824 -> String.format("%.3fMB", byteNum.toDouble() / 1048576)
            else -> String.format("%.3fGB", byteNum.toDouble() / 1073741824)
        }
    }

    private fun createDirectory(directoryPath: String) {
        val file = File(directoryPath)
        if (!file.exists()) {
            file.mkdirs()
        }
    }

    private fun directoryExists(directoryPath: String): Boolean {
        val file = File(directoryPath)
        return file.isDirectory
    }

    private fun copyFile(sourcePath: String, destinationPath: String) {
        var inChannel: FileChannel? = null
        var outChannel: FileChannel? = null
        try {
            val sFile = File(sourcePath)
            val dFile = File(destinationPath)

            inChannel = FileInputStream(sFile).channel
            outChannel = FileOutputStream(dFile).channel
            inChannel.transferTo(0, inChannel.size(), outChannel)

        } catch (e: IOException) {
        } finally {
            try {
                inChannel?.close()
                outChannel?.close()
            } catch (e: IOException) {
            }
        }
    }

    private fun deleteFile(filePath: String) {
        val file = File(filePath)
        if (file.exists()) {
            file.delete()
        }
    }

    private fun fileExists(filePath: String): Boolean {
        val file = File(filePath)
        return file.exists()
    }
}
